src/main/kotlin/no/skatteetaten/aurora/boober/service/RedeployService.kt

```diff
@@ -24,55 +28,82 @@ class RedeployService(val openShiftClient: OpenShiftClient, val openShiftObjectG
+        }
+
+        ImageStreamTag().from(imageStreamTagResponse).findErrorMessage()?.let {
+            return createFailedRedeployResult(it, imageStreamTagResponse)
+        }
+
+        val updatedImageStreamResponse = getUpdatedImageStream(deploymentSpec)
+                ?: return RedeployResult() // TODO return failed?
+        if (!updatedImageStreamResponse.success) {
+            return createFailedRedeployResult(updatedImageStreamResponse.exception, imageStreamTagResponse, updatedImageStreamResponse)
         }
 
-        val imageStreamImportResponse = importImageStream(deploymentSpec, redeployContext)
-                ?: return RedeployResult()
-        redeployContext.verifyResponse(imageStreamImportResponse).takeUnless { it.success }?.let {
-            return createFailedRedeployResult(it, imageStreamImportResponse)
+        val updatedImageStream = ImageStream().from(updatedImageStreamResponse)
+        updatedImageStream.findErrorMessage()?.let {
+            return createFailedRedeployResult(it, imageStreamTagResponse, updatedImageStreamResponse)
         }
 
-        if (redeployContext.didImportImage(imageStreamImportResponse)) {
-            return RedeployResult.fromOpenShiftResponses(listOf(imageStreamImportResponse))
+        if (updatedImageStream.isSameImage(imageStream)) {
+            val deploymentRequestResponse = performDeploymentRequest(deploymentSpec)
+            return RedeployResult.fromOpenShiftResponses(listOf(imageStreamTagResponse, updatedImageStreamResponse, deploymentRequestResponse))
         }
 
-        val deploymentRequestResponse = requestDeployment(deploymentSpec)
-        return RedeployResult.fromOpenShiftResponses(listOf(imageStreamImportResponse, deploymentRequestResponse))
+        return RedeployResult.fromOpenShiftResponses(listOf(imageStreamTagResponse, updatedImageStreamResponse))
     }
 
-    private fun createFailedRedeployResult(verificationResult: RedeployContext.VerificationResult, deploymentRequestResponse: OpenShiftResponse) =
-            RedeployResult(success = false, message = verificationResult.message, openShiftResponses = listOf(deploymentRequestResponse))
+    private fun createFailedRedeployResult(message: String?, vararg openShiftResponses: OpenShiftResponse) =
+            RedeployResult(success = false, message = message, openShiftResponses = openShiftResponses.toList())
 
-    private fun importImageStream(deploymentSpec: AuroraDeploymentSpec, redeployContext: RedeployContext): OpenShiftResponse? {
-        val imageStreamImportResource = generateImageStreamImportResource(redeployContext) ?: return null
-        val namespace = deploymentSpec.environment.namespace
-        val command = openShiftClient.createOpenShiftCommand(namespace, imageStreamImportResource)
+    private fun performImageStreamTag(namespace: String, imageStream: ImageStream): OpenShiftResponse? {
```

Comment:
Jeg synes denne skal returnere et svar uansett. Enten kan vi hive exceptions og flytte håndtering av dem ut, eller så bør vi returnere et svar med en ordentlig feilmelding. 

F.eks under her hvor du returnerer null kan du si "Cannot find image", eller "Cannot find image tag"

Reply:
Jeg begynner faktisk til å helle litt mot at man bør hive exceptions her og trekke ut feilhåndteringen. Da kan imageStream.findTagName også hive exception. For det er jo en feil som aldri skal skje. Hvis den skjer så er IS laget feil.

Reply:
Tenker du på at jeg kjører en try-catch i metoden som kaller performImageStreamTag?
Hva burde exception'en være i så fall, har vi en passende type til dette?

Reply:
Kanskje jeg kunne gjort noe sånt (men at RuntimException burde være en mer spesifikk type):

```
val imageStreamTagResponse = try {
  performImageStreamTag(deploymentSpec.environment.namespace, imageStream)
} catch (e: RuntimeException) {
  return createFailedRedeployResult(e.message)
}
```

Reply:
Hmm nå ble jeg jammen litt usikker her. Staten som vi sjekker for når vi returnerer null er jo noe som egentlig aldri skal skje her. 
Hva mener dere andre?